src/profile-logic/import/art-trace.js

```diff
@@ -0,0 +1,932 @@
+      validateMatchingVersions(version, summary.summaryVersion);
+      summaryDetails = summary.summaryDetails;
+      break;
+    } else {
+      throw new Error(`Invalid trace format: got invalid code ${code}.`);
+    }
+  }
+
+  // Now that we have the "clock" information from the summary section, read the
+  // records from earlier in the file.
+  const { clock } = summaryDetails;
+  const methodActions = [];
+  for (const recordStart of recordStartPositions) {
+    reader.setCurPos(recordStart);
+    methodActions.push(parseRecord(reader, version, recordSize, clock));
+  }
+
+  return {
+    summaryDetails,
+    threads,
+    methods,
+    methodActions,
+    startTimeInUsecSinceBoot,
+  };
+}
+
+function parseArtTrace(buffer: ArrayBuffer): ArtTrace {
+  const reader = new ByteReader(new Uint8Array(buffer));
+  switch (detectArtTraceFormat(buffer)) {
+    case 'regular':
+      return parseRegularFormat(reader);
+    case 'streaming':
+      return parseStreamingFormat(reader);
+    default:
+      throw new Error('Not an android trace');
+  }
+}
+
+// ART traces can be based on either sampling or method tracing:
+//  - Method tracing means that, every time any method is entered or exited,
+//    a timestamp is taken, and the enter/exit is recorded.
+//  - Sampling obtains stacks at a fixed sampling rate, usually based on wall
+//    clock time.
+// There is no property in the trace that lets us know which of the two was used
+// when this profile was obtained; the format makes both look like method
+// tracing. So we resort to measuring deltas between the timestamps on the
+// method actions: In sampling-based profiles, those timestamps are the original
+// sample timestamps. A sample that changes the current stack will cause a whole
+// bunch of methods to be exited and entered "simultaneously", i.e. with the
+// same timestamp.
+//
+// This function returns the average of the lowest 20% of timestamp deltas that
+// can be observed among the first 500 method actions.
+function procureSamplingInterval(trace) {
+  const { methodActions } = trace;
+
+  // Gather up to 500 time deltas between method actions on a thread.
+  const deltas: number[] = [];
+  const previousTimestampByThread = new Map();
+  const numberOfActionsToConsider = Math.min(500, methodActions.length);
+  for (let i = 0; i < numberOfActionsToConsider; i++) {
+    const { tid, globalTime } = methodActions[i];
+    const previousTimestamp = previousTimestampByThread.get(tid);
+    if (globalTime !== previousTimestamp) {
+      if (previousTimestamp !== undefined) {
+        const delta = globalTime - previousTimestamp;
+        deltas.push(delta);
+      }
+      previousTimestampByThread.set(tid, globalTime);
+    }
+  }
+
+  // Return the average of the lowest 20% of deltas.
+  deltas.sort((a, b) => a - b);
+  if (deltas.length < 5) {
+    return deltas[0];
+  }
+  const deltasThatMatter = deltas.slice(0, Math.floor(deltas.length / 5));
```

Comment:
Hmm... I wonder if this assumption could somehow.

Reply:
I'm sure it could somehow. More seriously, yeah, this isn't the prettiest code, but I think it doesn't have to be rock solid.